File: 046.py
#   046
#   Permutations
#   2015-07-06

#####################################################
class Solution:
    # @param {integer[]} nums
    # @return {integer[][]}
    def permute(self, nums):
        size = len(nums)
        v = [[nums[0]]]
        while len(v[0]) < size:
            a = v.pop(0)
            index = len(a)
            for i in range(index+1):
                c = list(a)
                c.insert(i, nums[index])
                v.append(c)
        return v

#####################################################

nums = [1,2,3]

sol = Solution()
print sol.permute(nums)

File: ruby/046.rb
# 046
# Permutations
# 2015-07-02
#####################################################
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
  v = [[nums.first]]
  while v.first.size < nums.size
    a = v.shift
    l = a.size
    0.upto(l) do |i|
      next if i < l and a[i] == nums[l]
      v.push(Array.new(a).insert(i, nums[l]))
    end
  end
  v
end

#####################################################

nums = [1,2,3]
puts nums.inspect
puts permute(nums).inspect

